### PR TITLE
[ESD-966] Sanitize parsing of setting string from SBP message

### PR DESCRIPTION
**Pull request by pmiettinen**
https://github.com/swift-nav/piksi_buildroot/pull/1046

### CODE REVIEW

File: src/settings_util.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        continue;
+      }
+      if (i + 1 < blen) {
+        *type = (const char *)&buf[i + 1];
+        break;
+      }
+
+    case 4:
+      /* Enum list sentinel NULL */
+      if (i == blen - 2) break;
```
Comment (pmiettinen) [inline src/settings_util.c]: Not sure if this NULL is supposed to be in the message but this is how it has been who knows how long..

File: src/settings_util.c [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    if (buf[i] != '\0') {
+      continue;
+    }
+    tok++;
+    switch (tok) {
+    case 1:
+      if (name == NULL) {
+        continue;
+      }
+      *name = (const char *)&buf[i + 1];
```
Comment (adel-mamin) [inline src/settings_util.c]: Maybe add `if (i + 1 < blen)` here as well.

Comment (pmiettinen) [reply src/settings_util.c]: Done.

File: src/settings.c [hunk 3]
```diff
@@ -384,11 +353,10 @@ static void settings_write_callback(uint16_t sender_id, uint8_t len, uint8_t *ms
   /* Extract parameters from message:
    * 3 null terminated strings: section, setting and value
```
Comment (yizow) [inline src/settings.c]: Should update this to new behaviour: potentially 4 null temrinated strings

File: src/settings.c [hunk 4]
```diff
@@ -384,11 +353,10 @@ static void settings_write_callback(uint16_t sender_id, uint8_t len, uint8_t *ms
...
    */
-  const char *section = NULL;
-  const char *name = NULL;
-  const char *value = NULL;
-  if (settings_parse((char *)msg, len, &section, &name, &value) != 0) {
-    ctx->api_impl.log(log_warning, "settings write message failed");
+  const char *section = NULL, *name = NULL, *value = NULL, *type = NULL;
+  int res = settings_parse((char *)msg, len, &section, &name, &value, &type);
+  if (res < 3) {
```
Comment (yizow) [inline src/settings.c]: I'm not sure what i think of `res < SETTINGS_TOKENS_VALUE`. Does it make sense for `res` to be "less than" `VALUE`? (In this case, implying that `VALUE` was missing)
Just a possibility

File: src/settings.c [hunk 5]
```diff
@@ -425,10 +393,8 @@ static int settings_update_watch_only(settings_t *ctx, char *msg, uint8_t len)
   /* Extract parameters from message:
    * 3 null terminated strings: section, setting and value
```
Comment (yizow) [inline src/settings.c]: Should update this to new behaviour: potentially 4 null temrinated strings

File: src/settings.c [hunk 6]
```diff
@@ -425,10 +393,8 @@ static int settings_update_watch_only(settings_t *ctx, char *msg, uint8_t len)
...
    */
-  const char *section = NULL;
-  const char *name = NULL;
-  const char *value = NULL;
-  if (settings_parse(msg, len, &section, &name, &value) != 0) {
+  const char *section = NULL, *name = NULL, *value = NULL, *type = NULL;
+  if (settings_parse(msg, len, &section, &name, &value, &type) < 3) {
```
Comment (yizow) [inline src/settings.c]: Same comment as above regarding `< 3`

File: src/settings_util.c [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  int tok = 0;
+  if (section) *section = (const char *)buf;
+  for (int i = 0; i < blen; i++) {
+    if (buf[i] != '\0') {
+      continue;
+    }
+    tok++;
+    switch (tok) {
+    case 1:
+      if (name == NULL) {
```
[the remark this answers is not in the document]
Comment (pmiettinen) [reply src/settings_util.c]: Done.

File: src/settings_util.c [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+      }
+
+    case 4:
+      /* Enum list sentinel NULL */
+      if (i == blen - 2) break;
+
+    case 5:
+      if (i == blen - 1) break;
+
+    default: return -1;
```
[the remark this answers is not in the document]
Comment (pmiettinen) [reply src/settings_util.c]: Done.

File: src/settings.c (hunk 3 above)
Comment (pmiettinen) [reply src/settings.c]: Done.

File: src/settings.c (hunk 4 above)
Comment (pmiettinen) [reply src/settings.c]: Not sure if I understood the review comment fully but I added further commenting and unified the styles for the two occurrences where `settings_parse` is called.

**Comment by pmiettinen** [comment]
@yizow The sentinel `NULL` is very confusing and makes the parsing unnecessarily complicated. I have tried to make it so that this change doesn't break existing settings tooling. As I stated in the buildroot side PR I think ultimately this `NULL` should be removed from the message but it should be done under a new issue and requires synchronizing between multiple repos.

File: src/settings_util.c (hunk 1 above)
Comment (pmiettinen) [reply src/settings_util.c]: See https://github.com/swift-nav/libsettings/pull/5#issuecomment-448558851

I'd say for now we just need to trust the sender is not malicious and straighten this issue later for good. @benjaminaltieri comments?

Comment (benjaminaltieri) [reply src/settings_util.c]: agreed, let's default to previous behaviors for this refactor and evaluate/update in separate PRs

**Pull request merged**